Support container provisioner in toolbox

thrix opened the pull request:
For Fedora Silverblue users it is common to run podman via `flatpak-spawn --host` which runs podman on the host system itself. This requires to pass the toolbox container name when running `podman cp` to correctly copy stuff from the toolbox container, where `tmt` is installed to the provisioned container.

Fixes #1020

Pull Request Checklist

* [x] implement the feature
* [x] extend the test coverage
* [x] include a release note

thrix commented:
@jkonecny12 this is the change I talked about, with it tmt container provisioner works well :)

happz commented on tmt/steps/provision/podman.py:
```diff
@@ -283,6 +284,33 @@ def _run_ansible(
+                if line.startswith('name="'):
+                    return line[6:-1]
+
+        return None
```
These two look like they belong to guest facts, as they are properties of a machine similar to `has_selinux` or `arch`, for example. Once moved, you would reach them via  `self.parent.plan.my_run.runner.facts.*`.

thrix replied:
thanks for the pointers

thrix replied:
moved to facts

happz commented on tmt/steps/provision/podman.py:
```diff
@@ -283,6 +284,33 @@ def _run_ansible(
+    @functools.cached_property
+    def _is_toolbox(self) -> bool:
+        """ Return ``True`` if running in toolbox, ``False`` otherwise. """
+        if os.path.exists('/run/.toolboxenv'):
```
`os.path` should be replaced with `Path().exists()`

thrix replied:
need to remember we use pathlib here :)

thrix replied:
moved to facts, where we run shell commands, so this does not apply anymore, resolving.

janhavlin commented:
I'm getting this error, any idea?
```
❯ tmt --version
tmt version: 1.37.0.dev20+gce8a3c6e

❯ tmt run -a provision -h container -i fedora:40 plan -n /testing-farm/sanity
/var/tmp/tmt/run-135

/testing-farm/sanity
    discover
        how: fmf
        directory: /var/home/jhavlin/work/tests
        tests: /testing-farm/script
        summary: 1 test selected
    provision
        queued provision.provision task #1: default-0
        
        provision.provision task #1: default-0
        how: container
        image: fedora:40
        fail: Command 'podman pull -q fedora:40' returned 1.
        fail: Command 'podman pull -q fedora:40' returned 1.
        fail: Command 'podman pull -q fedora:40' returned 1.
        fail: Command 'podman pull -q fedora:40' returned 1.
        fail: Command 'podman pull -q fedora:40' returned 1.
        fail: Retries of 'Pulling 'fedora:40' image' unsuccessful.
    finish
        warn: Nothing to finish, no guests provisioned.

plan failed

The exception was caused by 1 earlier exceptions

Cause number 1:

    provision step failed

    The exception was caused by 1 earlier exceptions

    Cause number 1:

        Retries of 'Pulling 'fedora:40' image' unsuccessful.

        The exception was caused by 5 earlier exceptions

        Cause number 1:

            Command 'podman pull -q fedora:40' returned 1.

            stderr (1/1 lines)
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
            Portal call failed: Failed to start command: Failed to change to directory “/var/tmp/tmt/run-135/testing-farm/sanity/provision/default-0” (No such file or directory)
            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
...
```

thrix commented:
@janhavlin this is something we should document, add this to your environment (I have it in .bashrc):

```
export TMT_WORKDIR_ROOT=$HOME/.local/share/tmt
```

janhavlin commented:
Thanks, I got it working!

jkonecny12 commented:
> @janhavlin this is something we should document, add this to your environment (I have it in .bashrc):
> 
> ```
> export TMT_WORKDIR_ROOT=$HOME/.local/share/tmt
> ```

Couldn't we have this resolved by default correctly? Requesting users to set somewhere something to make things work seems like an issue generator and unhappy users.

thrix commented:
> > @janhavlin this is something we should document, add this to your environment (I have it in .bashrc):
> > ```
> > export TMT_WORKDIR_ROOT=$HOME/.local/share/tmt
> > ```
> 
> Couldn't we have this resolved by default correctly? Requesting users to set somewhere something to make things work seems like an issue generator and unhappy users.

@jkonecny12 I agree it is sad, but we will need to have a discussion with `tmt` folks :)

thrix commented:
FTR the workdir autodetection will be another patch but it won't be in 1.40.

psss commented:
The newly added test failed because of a missing `toolbox` package. Should be fixed by e58860d0.

psss commented on tests/provision/container/toolbox/test.sh:
```diff
@@ -0,0 +1,159 @@
+            rlRun "type toolbox_run"
+
+            # Install all dependencies needed for hatch installation
+            rlRun "toolbox_run sudo dnf -y install git gcc hatch krb5-devel libpq-devel libvirt-devel python-devel"
```
What about using `make develop` instead of duplicating the list here?

thrix replied:
@psss yeah, I can try, is a problem to add to `make develop` also `hatch`?

happz replied:
`make develop` calls `make _deps` which installs `hatch python3-devel python3-hatch-vcs rpm-build`.

thrix replied:
@psss used `make develop`

psss commented on tests/provision/container/toolbox/test.sh:
```diff
@@ -0,0 +1,159 @@
+
+        # When running via tmt, we will use a new user to run toolbox
+        if env | grep -Eq "(PACKIT_COPR_PROJECT|TMT_TREE)"; then
+          rlRun "toolbox_user=toolbox" 0 "Generate toolbox container name"
```
Copy & paste comment?

thrix replied:
yes, fixed, thank you

psss reviewed (commented):
The change itself looks good. The test setup looks really super-complex and thus, I'm afraid, a bit fragile. Added a few comments though.

thrix commented:
> The newly added test failed because of a missing `toolbox` package. Should be fixed by [e58860d](https://github.com/teemtee/tmt/commit/e58860d0c402d6f528420d942e4ec3a9d1031a44).

I simplified, so I hope this concern is now addressed

thrix commented:
/packit build

psss reviewed (approved):
Looks much better now, thanks! Though it seems that tests are red. You've mentioned that `el9` should be disabled, right? Plus there's seems to be an [avc failure](https://artifacts.dev.testing-farm.io/b3d58468-a649-4881-8f8b-a5a12856a957/work-the-restd_1p16_m/plans/provision/container/the-rest/execute/data/guest/default-0/tests/provision/container/toolbox-15/checks/avc.txt) on `fedora-41+`.

thrix commented:
/packit build

thrix commented:
Issues here are unrelated, merging
